### ISSUE TITLE
`exportPreset` doesn't capture pane

### ISSUE THREAD

**Issue by FractalHQ** [cocopon/tweakpane#431]
I have a large root Pane with _many_ folders and parameters nested inside.  But when I call `pane.exportPreset()` on the root, it returns what appears to be a random folder from the middle of the tree.

<details>
<summary>screenshot</summary>

<img width="418" alt="Screenshot 2022-09-30 at 9 58 21 PM" src="https://user-images.githubusercontent.com/55504972/193378738-c0951ce7-9734-4d35-b165-e0349f36791b.png">

</details>

I'm not sure if this is intended functionality or not.  Is this because I neglected to specify the `presetKey` for each controller?  (I'm just now reading about `presetKey` 🤦‍♂️

As I try to figure it out, it appears there is no API exposed at all for things like recursive children, controller values / titles / etc.  

<details>
<summary>I got the folders recursively in the screenshot like so:</summary>

```ts
const walk = (gui: FolderApi) => {
	for (const child of gui.children) {
		console.log(child)
		const isFolder = child.rackApi_ !== undefined
		if (isFolder) {
			console.log('🗂')
			walk(child as FolderApi)
		}
	}
}
```

</details>

But now, getting all the targets / titles / values doesn't seem to be very straightforward.

Unless I'm missing something big, it seems like my best bet is to build a system over top of tweakpane to expose important information like values, names, targets, default values, etc in order to implement a preset system that functions like [`gui.save()` in `lil-gui`](https://lil-gui.georgealways.com/#Guide#Saving) or `dat.gui`

Sorry for the long post!  I'm knee deep in a huge project and was tasked with implementing presets so I'm a bit scatter brained 😅

**Comment by cocopon** [comment]
As you said, there is no way to export a setting of all blades for now. I designed `exportPreset` for exporting all bindings, not blades. But I know that many users want to export all settings, so I'll consider it in the next major version.

**Comment by cocopon** [comment]
There are several possibilities for the preset format.

![image](https://user-images.githubusercontent.com/602961/219870844-06db5d34-1f6e-41cc-bb27-ea963933fa75.png)

1. Flatten object (current format)

```js
{
  size: 16,
  weight: 'Normal',
  name: 'Sketch',
  active: true,
  color: '#ff0055',
  offset: {x: 0, y: 0},
  point3d: {x: 0, y: 0, z: 0},
  point4d: {x: 0, y: 0, z: 0, w: 0},
}
```

**Pros:**
- Looks clean.
- Not depend on the UI structure.

**Cons:**
- The output can be ambiguous if the pane contains bindings with duplicate keys.
  - `presetKey` can solve this problem, but not smart a bit...

2. Nested object

```js
{
  size: 16,
  weight: 'Normal',
  name: 'Sketch',
  active: true,
  color: '#ff0055',
  tab: {
    offset: {x: 0, y: 0},
    point3d: {x: 0, y: 0, z: 0},
    point4d: {x: 0, y: 0, z: 0, w: 0},
  },
}
```

**Pros:**
- Easy to understand the structure.

**Cons:**
- The output can also be ambiguous.
- How to decide a container name like tabs...?
- Exported object depends on the UI structure. If user changes the structure, importing the preset can be failed.

3. Nested array

```js
[
  {key: 'size', value: 16},
  {key: 'weight', value: 'Normal'},
  {key: 'name', value: 'Sketch'},
  {key: 'active', value: true},
  {key: 'color', value: '#ff0055'},
  // tab
  [
    [
      {key: 'offset', value: {x: 0, y: 0}},
    ],
    [
      {key: 'point3d', value: {x: 0, y: 0, z: 0}},
      {key: 'point4d', value: {x: 0, y: 0, z: 0, w: 0}},
    ],
  ],
]
```

**Pros:**
- The output is simple and unique.
- User can convert it to the current format with the small utility function.

**Cons:**
- Exported object depends on the UI structure. If user changes the structure, importing the preset can be failed.

**Comment by FractalHQ** [comment]
Hmmm... this is a tricky one.  I think the most important thing is to expose the data of each `Blade` and make it easily accessible.  That way, no matter which way you decide to go with your custom preset system, developers can still create their own system if the built-in one doesn't fit their needs.

Simply being able to do `Blade.toJSON()` or `Blade.get()` would be a catch-all primitive that enables freedom and flexibilty.  Then, building a preset system on top like you describe wouldn't affect a developers ability to manage the data how they need.

I'll have to think a bit more about these 3 options in the meantime.

**Comment by FractalHQ** [comment]
For example, this audio-reactivity management system has quite unique requirements that I don't think could be solved by 1, 2, or 3:

<details>
<summary>Screenshot</summary>

<img width="512" alt="Screenshot 2023-02-18 at 2 04 43 PM" src="https://user-images.githubusercontent.com/55504972/219883663-b76ea135-1152-431a-8288-9937b3155443.png">

</details>

And other `FolderAPI` instances have nested `TabApi` instances to manage presets for each folder:

<details>
<summary>Screenshot</summary>

<img width="533" alt="Screenshot 2023-02-18 at 2 05 16 PM" src="https://user-images.githubusercontent.com/55504972/219883738-62b15f33-eb25-41e9-9daa-1c9cf0790ca5.png">

</details>

I think the best solution would be to have `Blade.get()` to get the data for custom systems like mine.

For simpler / common use cases, I like option 2:

**2. Nested object**

> - How to decide a container name like tabs...?

I think all `Blades` (as well as `Folder` and `Tab`) should require a `presetKey` when using a preset system.  For key defaults, perhaps they could use the readonly name of the blade + key or index (i.e. `color_background` and for tabs, just the index so `tab_0` and `tab_1`).

> - Exported object depends on the UI structure. If user changes the structure, importing the preset can be failed.
We could use the preset-keys, and provide a warning when a preset object isn't able to be loaded.  I think this is OK, because it's up to the developer to establish and manage their data structures when using preset systems.

**Comment by cocopon** [comment]
Thank you for the detail! I'm amazed at your heavily customzed pane.

>  think the most important thing is to expose the data of each Blade and make it easily accessible. That way, no matter which way you decide to go with your custom preset system, developers can still create their own system if the built-in one doesn't fit their needs.

It makes sense. I'll try to implement this functionality...

**Comment by FractalHQ** [comment]
Haha thanks, I love hacking on tweakpane! 🙏  Let me know if there is anything I can do to help.

**Comment by cocopon** [comment]
After considering, preset will be replaced with "blade state" in v4.

![image](https://user-images.githubusercontent.com/602961/223585429-eb3a2aff-0b9b-4b1a-aef1-f1ae15377b1b.png)

```js
// v3: preset
{
  "name": "exported json",
  "size": 10,
  "color": "#00ffd6",
}
```

```js
// v4: blade state
{
  "disabled": false,
  "hidden": false,
  "children": [
    {
      "disabled": false,
      "hidden": false,
      "label": "name",
      "value": "exported json",
      "key": "name"
    },
    {
      "disabled": false,
      "hidden": false,
      "label": "size",
      "value": 10,
      "max": 100,
      "min": 0,
      "key": "size"
    },
    {
      "disabled": false,
      "hidden": false,
      "label": "color",
      "value": "#00ffd6",
      "key": "color"
    }
  ],
  "expanded": true,
  "title": "Values"
}
```

Every blade has the method `importState()` and `exportState()`, and the pane uses them recursively to create a whole state.

Blade state contains more information about blades and it can convert into classic preset easily with small utility function.

**Comment by ashthornton** [comment]
Big +1 for this feature! I've actually recently ended up hacking together my own version of this which works nicely but required overwriting core methods so it isn't future-proofed.

What's nice about the ability to export all data is that you can place this in localStorage and retrieve it on refresh so you can persist your settings. It might be nice to have this as a built-in option on tweakpane :)

I also found that I needed the ability to exclude certain folders from the export for things like monitor bindings and other settings since there was no use for them when re-importing.

**Comment by cocopon** [comment]
You'll be able to except monitor bindings by `readonly: true`.

**Comment by ashthornton** [comment]
> > I also found that I needed the ability to exclude certain folders from the export for things like monitor bindings and other settings since there was no use for them when re-importing.
> 
> You'll be able to except monitor bindings by `readonly: true`.

Would this also work for other blades/folders?

**Comment by cocopon** [comment]
After some thoughts, I'll update the structure of the binding blade state from:

```json
{
  "key": "foo",
  "value": 1,

  "disabled": false,
  "hidden": false,
  "label": "foo"
}
```

to:

```json
{
  "binding": {
    "key": "foo",
    "value": 1
  },

  "disabled": false,
  "hidden": false,
  "label": "foo"
}
```

Because `key` and `value` is too general. You'll be able to determine binding blades easily than ever.

**Comment by FractalHQ** [comment]
Sweet!

Would an additional generic `data` field make sense here for arbitrary user data?

**Comment by cocopon** [comment]
> Would an additional generic data field make sense here for arbitrary user data?

Blade state is literally a state of the blade, so I think related data is out of scope of the state. All binding blades have a string field `tag` and you can use it as a key for the data like this:

```js
const dataMap = {
  foo: data1,
  bar: data1,
};

pane.addInput(PARAMS, 'param1', {tag: 'foo'});
pane.addInput(PARAMS, 'param2', {tag: 'bar'});
```

Blade state:
```json
{
  "disabled": false,
  "hidden": false,
  "children": [
    {
      "disabled": false,
      "hidden": false,
      "label": "param1",
      "binding": {
        "key": "param1",
        "value": 1
      },
      "tag": "foo"
    },
    {
      "disabled": false,
      "hidden": false,
      "label": "param2",
      "binding": {
        "key": "param2",
        "value": 2
      },
      "tag": "bar"
    }
  ],
  "expanded": true
}
```

**Comment by autonome** [comment]
Glad to see this feature being added!

Until v4 is released, did anyone figure out a workaround for exporting the state of a pane?

**Comment by cocopon** [comment]
Released in v4.0.0.
